### PR TITLE
refactor: Coinjoin Status fetch schedule based on deadline 

### PR DESCRIPTION
**Pull request by szymonlesisz**
## Description
active `CoinjoinRound.phaseDeadline` is followed by `Status`. Status is scheduling own fetch timeout using following logic:

1. phaseDeadline < STATUS_TIMEOUT.enabled / 2 ? `timeout = STATUS_TIMEOUT.enabled / 2`
2. phaseDeadline >  STATUS_TIMEOUT.enabled / 2 && phaseDeadline < STATUS_TIMEOUT.enabled ?  `timeout = phaseDeadline`
3. phaseDeadline >  STATUS_TIMEOUT.enabled?  `timeout = STATUS_TIMEOUT.enabled`

## Commits

1. add shared Status.log function and fix missing `changed` event on unexpected phase change
2. add Status following round logic, refactor Status tests and use `jest fake timers` which is not only more reliable but also reduces execution time from ~43 sec to ~3 sec(!!!)
3. fix round deadline estimation for round with phase `Ended`

### CODE REVIEW

File: packages/coinjoin/src/client/Status.ts [hunk 1]
```diff
@@ -110,15 +140,23 @@ export class Status extends EventEmitter {
...
-        const nearestDeadline = findNearestDeadline(this.rounds);
-        // TODO: add timeout randomness?
+        const defaultTimeout = STATUS_TIMEOUT[this.mode];
+        const half = defaultTimeout / 2;
+
+        // get deadlines from registered rounds
+        // set minimum deadline to half of defaultTimeout to avoid http request overflow
+        const nearestDeadlines = this.registeredRound
+            .map(r => r.phaseDeadline - Date.now())
+            .map(r => (r > half ? r : half));
```
Comment (matejkriz) [inline packages/coinjoin/src/client/Status.ts]: nitpick, slightly easier to read?  
```suggestion
            .map(r => Math.max(r, half));
```

File: packages/coinjoin/tests/client/Status.test.ts [hunk 2]
```diff
@@ -1,115 +1,204 @@
...
-    AFFILIATE_INFO,
-    DEFAULT_ROUND,
-    STATUS_EVENT,
-    ROUND_CREATION_EVENT,
-} from '../fixtures/round.fixture';
+import { AFFILIATE_INFO, DEFAULT_ROUND, STATUS_EVENT } from '../fixtures/round.fixture';
 
-let server: Awaited<ReturnType<typeof createServer>>;
+// using fakeTimers and async callbacks
+const fastForward = (time: number) => {
```
Comment (matejkriz) [inline packages/coinjoin/tests/client/Status.test.ts:8]: nice 👍

**Review by matejkriz — APPROVED** [review]
Looks good, nice usage of the fake timers. Feel free to ignore the nitpick and merge...

File: packages/coinjoin/src/client/Status.ts (hunk 1 above)
Comment (szymonlesisz) [reply packages/coinjoin/src/client/Status.ts]: sure, fixed rebased and forcepushed

**Pull request merged**